msal_extensions/token_cache.py

```diff
@@ -1,156 +1,113 @@
-        super(WindowsTokenCache, self).__init__(cache_location, **kwargs)
-        self._dp_agent = WindowsDataProtectionAgent(entropy=entropy)
-
-    def _write(self, contents):
-        with open(self._cache_location, 'wb') as handle:
-            handle.write(self._dp_agent.protect(contents))
-
-    def _read(self):
-        with open(self._cache_location, 'rb') as handle:
-            cipher_text = handle.read()
-        return self._dp_agent.unprotect(cipher_text)
 
+def build_token_cache(  # pylint: disable=too-many-arguments
+        location,
+        entropy='',  # For Windows
+        service_name="", account_name="",  # For OSX
+        schema_name="", attributes=None, fallback_to_plaintext=False,  # For Linux
+        **kwargs):  # Extra parameters will be relayed to LibsecretPersistence
+    """Build a token cache which uses native encryption mechanism on local platform.
 
-class OSXTokenCache(FileTokenCache):
-    """A SerializableTokenCache implementation which uses native Keychain libraries to protect your
-    tokens.
+    It will use DPAPI when running on Windows, Keychain when running on OSX,
+    libsecret when running on Linux and optionally fallback to plain texxt file.
     """
-
+    persistence = build_persistence(
+        location,
+        entropy=entropy,
+        service_name=service_name,
+        account_name=account_name,
+        schema_name=schema_name,
+        attributes=attributes or {"msal_token_cache": "1.0"},
+        fallback_to_plaintext=fallback_to_plaintext,
+        dummy="{}",  # A valid json payload can be read by FilePersistence()
+        **kwargs)  # Extra parameters will be relayed to LibsecretPersistence
+    return PersistedTokenCache(persistence)
+
+
+class FileTokenCache(PersistedTokenCache):
+    """A token cache which uses plain text file to store your tokens."""
+    def __init__(self, cache_location, **ignored):  # pylint: disable=unused-argument
+        warnings.warn("You are using an unprotected token cache", RuntimeWarning)
+        warnings.warn("Use build_token_cache(...) instead", DeprecationWarning)
+        super(FileTokenCache, self).__init__(FilePersistence(cache_location))
+
+UnencryptedTokenCache = FileTokenCache  # For backward compatibility
+
+
+class WindowsTokenCache(PersistedTokenCache):
+    """A token cache which uses Windows DPAPI to encrypt your tokens."""
+    def __init__(
+            self, cache_location, entropy='',
+            **ignored):  # pylint: disable=unused-argument
+        warnings.warn("Use build_token_cache(...) instead", DeprecationWarning)
+        super(WindowsTokenCache, self).__init__(
+            FilePersistenceWithDataProtection(cache_location, entropy=entropy))
+
+
+class OSXTokenCache(PersistedTokenCache):
+    """A token cache which uses native Keychain libraries to encrypt your tokens."""
     def __init__(self,
                  cache_location,
                  service_name='Microsoft.Developer.IdentityService',
```

Comment:
This default value is not defined in the .NET or Java extension - though it is exactly where Visual Studio for Mac stores the token cache. I'm fine with having a default value, though just want to point it out in case anyone else has an opinion.

Reply:
Good observation, @jianghaolu , even on something NOT being changed, thus not highlighted in this PR! This is indeed a good topic to raise. Here come my thoughts.

* The current default values were pre-existing before this PR. I was not sure where/how they came from. I suspect it matches some values other Microsoft-owned products chose to use. Thanks for your confirmation.
* However, I would actually prefer that all our MSAL EX to NOT provide such default value here. Because this MSAL EX will also be used by 3rd-party app developers. And 1st-party and 3rd-party apps might not want to share tokens with each other.
* In fact, this PR intends to deprecate this entire `OSXTokenCache` class. And the new `build_token_cache(...)` factory method uses a set of non-Microsoft settings as default values. All apps are expected to provide their own settings.

Reply:
It makes sense to keep it backward compatible in here and require full settings from the new configuration API.